File: src/java/dsa/MyList.java
package dsa;

/**
 * Interface which defines methods for lists.
 *
 * @param <T> data type
 */
public interface MyList<T> {
    /**
     * Inserts an element at the specified index.
     *
     * @param index index to add element
     * @param element element to add
     * @throws IndexOutOfBoundsException if index is out of bounds (index must be between 0 and size(), inclusive)
     */
    public void add(int index, T element) throws IndexOutOfBoundsException;

    /**
     * Inserts an element at the end of this list.
     *
     * @param element element to add
     */
    public void add(T element);

    /**
     * Empties this list of all elements.
     */
    public void clear();

    /**
     * Checks if an element is in this list.
     *
     * @param element element to check for
     * @return true if found, false otherwise
     */
    public boolean contains(T element);

    /**
     * Compares an object with this list for equality.
     *
     * @param object object to compare to this list
     * @return true if object and this list are equal
     */
    @Override
    public boolean equals(Object object);

    /**
     * Retrieves, but does not remove, an element from this list at the specified index.
     *
     * @param index index to retrieve element from
     * @return element found
     * @throws IndexOutOfBoundsException if index is out of bounds (index must be between 0 and size() - 1, inclusive)
     */
    public T get(int index) throws IndexOutOfBoundsException;

    /**
     * Returns the index of the first occurrence of an element found in this list.
     *
     * @param element element to search for
     * @return index of the first occurrence of element, or -1 if not found
     */
    public int indexOf(T element);

    /**
     * Checks if this list has no elements.
     *
     * @return true if this list is empty, false otherwise
     */
    public boolean isEmpty();

    /**
     * Returns the index of the last occurrence of an element found in this list.
     *
     * @param element element to search for
     * @return index of the last occurrence of element, or -1 if not found
     */
    public int lastIndexOf(T element);

    /**
     * Retrieves and removes an element from this list at a specified index.
     *
     * @param index index to remove element from
     * @return element found at index
     * @throws IndexOutOfBoundsException if index is out of bounds (index must be between 0 and size() - 1, inclusive)
     */
    public T remove(int index) throws IndexOutOfBoundsException;

    /**
     * Retrieves and removes the first occurrence of an element from this list.
     *
     * @param element element to remove first occurrence of
     * @return true if successful, false otherwise
     */
    public boolean removeElement(T element);

    /**
     * Sets the existing element at a specified index to a new value.
     *
     * @param index index to set element at
     * @param element new value to set existing element to
     * @return old value of the element at position index
     * @throws IndexOutOfBoundsException if index is out of bounds (index must be between 0 and size() - 1, inclusive)
     */
    public T set(int index, T element) throws IndexOutOfBoundsException;

    /**
     * Returns the number of elements in this list.
     *
     * @return size of list
     */
    public int size();

    /**
     * Returns an array containing all the elements in this list.
     *
     * @return array of list elements
     */
    public Object[] toArray();

    /**
     * Returns a string representation of this list, e.g. "[element1, element2, element3, ..., elementN]".
     *
     * @return string form of this list
     */
    @Override
    public String toString();
}

File: src/go/dsa/mylist.go
package dsa

//go:generate mockgen -source=mylist.go -destination=mock_mylist.go -package=dsa

// MyList is an interface which defines methods for lists.
type MyList[T comparable] interface {
	// Add inserts the given element at the given index of a list. Returns an error if index
	// is out of bounds (index must be between 0 and Size(), inclusive).
	Add(index uint, element T) error

	// AddToEnd inserts the given element at the end of a list.
	AddToEnd(element T)

	// Clear empties a list of all elements.
	Clear()

	// Contains reports whether the given element is in a list.
	Contains(element T) bool

	// Equals reports whether obj is equal to a list.
	Equals(obj any) bool

	// Get returns, but does not remove, an element from a list at the given index.
	// Returns an error if index is out of bounds (index must be between 0 and
	// Size() - 1, inclusive).
	Get(index uint) (T, error)

	// IndexOf returns the index of the first occurrence of the given element found
	// in a list, or -1 if not found.
	IndexOf(element T) int

	// IsEmpty reports whether a list has no elements.
	IsEmpty() bool

	// IndexOf returns the index of the last occurrence of the given element found
	// in a list, or -1 if not found.
	LastIndexOf(element T) int

	// Remove retrieves and removes an element from a list at the given index.
	// Returns an error if index is out of bounds (index must be between 0 and
	// Size() - 1, inclusive).
	Remove(index uint) (T, error)

	// RemoveElement attempts to remove the given element from a list, and reports
	// whether the deletion was successful.
	RemoveElement(element T) bool

	// Set sets the existing element at the given index of a list to the given element and
	// returns the existing element. Returns an error if index is out of bounds
	// (index must be between 0 and Size() - 1, inclusive).
	Set(index uint, element T) (T, error)

	// Size returns the number of elements in a list.
	Size() uint

	// ToArray returns an array containing all the elements in a list.
	ToArray() []T

	// ToString returns a string representation of a list, e.g.
	// "[element1, element2, element3, ..., elementN]".
	ToString() string
}
